### PR TITLE
Easy 1328

### PR DESCRIPTION
**Pull request by alexal14**
Fixes EASY- it downloads a single file of a specified bag or a zipped bag

#### When applied it will...
* 
* 
* 

#### Related pull requests on github

repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.bagstore/component/BagStoresComponent.scala [hunk 1]
```diff
@@ -15,16 +15,16 @@
...
+import java.nio.file.{Files, Path}
 import java.util.UUID
 
 import nl.knaw.dans.easy.bagstore._
 import nl.knaw.dans.lib.error._
 import org.apache.commons.io.FileUtils
 
 import scala.language.postfixOps
-import scala.util.{ Failure, Success, Try }
+import scala.util.{Failure, Success, Try}
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.bagstore/component/BagStoresComponent.scala]: Applying the DANS code formatter would prevent this change.

File: src/main/scala/nl.knaw.dans.easy.bagstore/component/BagStoreComponent.scala [hunk 2]
```diff
@@ -81,7 +81,7 @@ trait BagStoreComponent {
               location <- fileSystem.toLocation(bagId)
               stagingDir <- processor.stageBagDir(location)
               stagedBag = stagingDir.resolve(location.getFileName)
-              _ <- processor.complete(stagedBag)
+                _ <- processor.complete(stagedBag)
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.bagstore/component/BagStoreComponent.scala:84]: please undo this chance

File: src/main/scala/nl.knaw.dans.easy.bagstore/component/BagStoresComponent.scala [hunk 3]
```diff
@@ -50,6 +50,19 @@ trait BagStoresComponent {
         }
     }
 
+    def getStream(itemId: ItemId, output: => OutputStream, fromStore: Option[Path] = None): Try[Unit] = {
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.bagstore/component/BagStoresComponent.scala:53]: The line above is the only difference with the `get` method. I don't understand the magic. Do you really need to return a Unit? At first sight I'd expect a stream.

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/BagsServletComponent.scala [hunk 4]
```diff
@@ -42,16 +45,26 @@ trait BagsServletComponent extends DebugEnhancedLogging {
...
         }
-        .flatMap(_.toBagId)
-        .flatMap(bagStores.enumFiles(_))
-        .map(bagIds => Ok(bagIds.mkString("\n")))
+        .flatMap{
+          case bagId: BagId =>
+            debug(s"Retrieving item $bagId")
+            request.getHeader("Accept") match {
+              case "application/zip" => bagStores.getStream(bagId, response.outputStream).map(_ => Ok())
+              case "text/plain" | "*/*" | null => bagStores.enumFiles(bagId).map(files => Ok(files.toList.mkString("\n")))
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.bagstore/server/BagsServletComponent.scala:61]: An empty OK doesn't seem of much use to me, we'll need the stream or its content.

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/BagsServletComponent.scala [hunk 5]
```diff
@@ -60,5 +73,31 @@ trait BagsServletComponent extends DebugEnhancedLogging {
...
+        case Seq(path) =>
+          ItemId.fromString(s"""$uuidStr/${ path }""")
+            .recoverWith {
+              case _: IllegalArgumentException => Failure(new IllegalArgumentException(s"invalid UUID string: $uuidStr"))
+            }
+            .flatMap(itemId => {
+              debug(s"Retrieving item $itemId")
+              bagStores.getStream(itemId, response.outputStream)
+            })
+            .map(_ => Ok())
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.bagstore/server/BagsServletComponent.scala:89]: again I would not expect an empty OK()

**Review by jo-pol — CHANGES_REQUESTED** [review]

File: src/main/scala/nl.knaw.dans.easy.bagstore/server/BagsServletComponent.scala [hunk 6]
```diff
@@ -42,16 +45,26 @@ trait BagsServletComponent extends DebugEnhancedLogging {
         })
     }
 
+
     get("/:uuid") {
```
Comment (PaulBoon) [inline src/main/scala/nl.knaw.dans.easy.bagstore/server/BagsServletComponent.scala:49]: We decided not to have zipping functionality for PDBS, so please get it out of this PR. We can do zipping later on, maybe even in another module.

**Comment by PaulBoon** [comment]
I will clean things up and do another PR for this.

**Comment by lindareijnhoudt** [comment]
this is part of
https://github.com/DANS-KNAW/easy-bag-store/pull/43
so i am closing this one

**Pull request closed without merging**